### ISSUE TITLE
How about react-map-gl's 'controlled viewport state' strategy?

### ISSUE THREAD

**Issue by janbaykara** [alex3165/react-mapbox-gl#572]
I've used react-mapbox-gl profitably for a while, and it's great.

However there are often issues with updating the viewport. React-map-gl (Uber) has a really robust strategy of being viewport-stateless. The viewport state is hosted by a component, and user interactions fire a callback that can be used to update the state.

I think this is a really powerful and simple strategy. It means that you can affect one-off changes to the viewport state that can then be changed by user interaction. At the moment this isn't so possible.

What do we think of emulating react-map-gl?

**Comment by janbaykara** [comment]
For now I'm welding the two together. I use `react-map-gl`'s `<InteractiveMap> to handle camera state, and pass the mapbox `map` instance to `react-mapbox-gl` layer components through context, as it expects.

```
class HybridMap {
    state = {
        viewport: {
            latitude: 40,
            longitude: -100,
            zoom: 3,
            bearing: 0,
            pitch: 0,
            width: 500,
            height: 500
        }
    }

    _getMap = (map) => {
        if (!map) return
        this.map = map.getMap()
    }

    getChildContext() {
        if (!this.map) return
        return { map: this.map }
    }

    static childContextTypes = {
        map: PropTypes.object
    }

    _onViewportChange = viewport => this.setState({ viewport })

    render () {
        return (
            <MapGL
                {...viewport}
                reuseMaps
                ref={this._getMap}
                onViewportChange={this._onViewportChange}
                mapboxApiAccessToken={MAPBOX_TOKEN}>

                <GeoJSONLayer
                    data={someData}
                    lineLayout={{}}
                    linePaint={{}}
                />

            </MapGL>
        )
    }
}

```

**Comment by brendanmoore** [comment]
I personally agree with this type of approach. It is much more "react" I think the `zoom, center, pitch bearing` as arrays are the source of a lot of user issues reported here (I see many arrays created in render) It will be a significant breaking change, but I would be keen.

**Comment by guyazulay** [comment]
Hi @janbaykara , 
I'm interested in your approach , can you please explain more about the issue of passing the instance through the context ? 
(it's not so clear in your example)

**Comment by janbaykara** [comment]
Hi @guyazulay, it's possible the specific code I provided is no longer valid since it looks like `react-mapbox-gl` now uses the new Context API.

The abstract strategy is to pass `react-map-gl`'s map instance through to `react-mapbox-gl` somehow. In my example, I manually set the context. I guess you could achieve a similar thing now by doing:
```
import { MapContext } from 'react-mapbox-gl/context`
<MapContext.Provider value={mapFromReactMapGL}>...</>
```

**Comment by guyazulay** [comment]
@janbaykara  how do you import the MapContext ? its not exportable in react-mapbox-gl .

**Comment by krisaoe** [comment]
@guyazulay MapContext is a named export from react-mapbox-gl/lib/context
you can import like so:
```js
import { MapContext } from 'react-mapbox-gl/lib/context';
```

**Comment by guyazulay** [comment]
yes, i already did it, thanks. However , it seems that it doesn't really work (i'm not sure why).
 Meanwhile, i passed the map directly as a prop to the component

**Comment by yesuanzao** [comment]
I suggest adding uncontrolled props like 'defaultZoom', 'defaultCenter', etc.

**Comment by mklopets** [comment]
btw, `MapContext` shouldn't be imported from `/lib/context` any more – it should be available from the main entrypoint
